dask/array/slicing.py

```diff
@@ -835,6 +835,13 @@ def check_index(ind, dimension):
 def slice_with_dask_array(x, index):
     from .core import Array, atop, elemwise
 
+    if all([len(index) == 1,
+            isinstance(index[0], Array),
+            index[0].dtype == bool,
+            index[0].ndim == x.ndim]):
+        x = x.ravel()
+        index = tuple(i.ravel() for i in index)
```

Comment:
I wonder if, instead of raveling, we should reshape as many dimensions as there are in the indexer.  This would be useful in cases like the following:

```python
In [1]: import numpy as np

In [2]: x = np.arange(48).reshape((4, 4, 3))

In [3]: y = np.arange(16).reshape((4, 4))

In [4]: x[y % 2 == 0]
Out[4]: 
array([[ 0,  1,  2],
       [ 6,  7,  8],
       [12, 13, 14],
       [18, 19, 20],
       [24, 25, 26],
       [30, 31, 32],
       [36, 37, 38],
       [42, 43, 44]])
```

Reply:
Was thinking about that as well. However was unable to produce that behavior with Dask Arrays as it appears to be unsupported currently. Suspect this should be a simple reshape of `x` when we do address those cases. Though wouldn't want to unintentionally lock us into the wrong behavior there. Maybe we could check that the shapes are equal instead of just ndim to guard against this applying in other cases until we figure them out?

Reply:
Yeah, I would check for equality of dimension (probably accepting nan == nan in this case) in the leading dimensions.  If so then I would reshape the first matching dimensions to be a singleton dimension.  Then I would call atop.  I *think* that this would work, but haven't thought particularly hard about it.

Reply:
Yeah, wasn't sure if `ravel` would work for `nan` dimensions. So played with it locally and it seems that fails ( https://github.com/dask/dask/issues/3194 ). Might need to think more about how to fix this as it would be weird if this matches NumPy for Dask Arrays with known dimensionality, but doesn't for unknown dimensionality.